consensus/hotstuff/forks/forks.go

```diff
@@ -225,34 +259,62 @@ func (f *Forks) checkForDoubleProposal(container *BlockContainer) {
 	}
 }
 
-// updateConsensusState updates consensus state.
+// updateFinalizedBlockQC updates the latest finalized block, if possible.
+// This function should be called every time a new block is added to Forks.
+// If the new block is the head of a 2-chain satisfying the finalization rule,
+// then we update Forks.lastFinalizedBlockQC to the new latest finalized block.
 // Calling this method with previously-processed blocks leaves the consensus state invariant.
 // UNVALIDATED: assumes that relevant block properties are consistent with previous blocks
-func (f *Forks) updateConsensusState(blockContainer *BlockContainer) error {
-	ancestryChain, err := f.getThreeChain(blockContainer)
-	// We expect that getThreeChain might error with a ErrorPrunedAncestry. This error indicates that the
-	// 3-chain of this block reaches _beyond_ the last finalized block. It is straight forward to show:
-	// Lemma: Let B be a block whose 3-chain reaches beyond the last finalized block
+// Error returns:
+// * model.ByzantineThresholdExceededError if we are finalizing a block which is invalid to finalize.
+//   This either indicates a critical internal bug / data corruption, or that the network Byzantine
+//   threshold was exceeded, breaking the safety guarantees of HotStuff.
+// * generic error in case of unexpected bug or internal state corruption
+func (f *Forks) updateFinalizedBlockQC(blockContainer *BlockContainer) error {
+	ancestryChain, err := f.getTwoChain(blockContainer)
+	// We expect that getTwoChain might error with a ErrPrunedAncestry. This error indicates that the
+	// 2-chain of this block reaches _beyond_ the last finalized block. It is straight forward to show:
+	// Lemma: Let B be a block whose 2-chain reaches beyond the last finalized block
 	//        => B will not update the locked or finalized block
-	if errors.Is(err, ErrPrunedAncestry) { // blockContainer's 3-chain reaches beyond the last finalized block
+	if errors.Is(err, ErrPrunedAncestry) {
+		// blockContainer's 2-chain reaches beyond the last finalized block
 		// based on Lemma from above, we can skip attempting to update locked or finalized block
 		return nil
 	}
-	if err != nil { // otherwise, there is an unknown error that we need to escalate to the higher-level application logic
-		return fmt.Errorf("retrieving 3-chain ancestry failed: %w", err)
+	if model.IsMissingBlockError(err) {
```

Comment:
would suggest to move this into the [`if err != nil` block](https://github.com/onflow/flow-go/blob/5ce21f7803465beeba156fa7052abd67f3029e82/consensus/hotstuff/forks/forks.go#L288) to reduce repeated error checks on the hot path